### PR TITLE
Updates NTP's Sponsored Images toggle functionality.

### PR DESCRIPTION
**Pull request by mkarolin**
Resolves https://github.com/brave/brave-browser/issues/27304

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by sangwoo108 — APPROVED** [review]
This looks good to me! But you might want to wait for reviews from @petemill or @fallaciousreasoning

File: components/brave_new_tab_ui/containers/newTab/settings/sponsoredImagesToggle.tsx
Comment (fallaciousreasoning) [inline components/brave_new_tab_ui/containers/newTab/settings/sponsoredImagesToggle.tsx:160]: Nit: I think functions should be lower case (except for components) to fit in with most other stuff. This is pretty inconsistently enforced though (sorry)

```suggestion
function getDescriptionText (rewardsEnabled: boolean,
```

File: components/brave_new_tab_ui/containers/newTab/settings/sponsoredImagesToggle.tsx
Comment (fallaciousreasoning) [inline components/brave_new_tab_ui/containers/newTab/settings/sponsoredImagesToggle.tsx:150]: Nit: I think functions should be lower case (except for components) to fit in with most other stuff. This is pretty inconsistently enforced though (sorry)

```suggestion
function getInfoTooltipText (checked: boolean, rewardsEnabled: boolean) {
```

File: components/brave_new_tab_ui/containers/newTab/settings/sponsoredImagesToggle.tsx
Comment (fallaciousreasoning) [inline components/brave_new_tab_ui/containers/newTab/settings/sponsoredImagesToggle.tsx:173]: Nit: I think functions should be lower case (except for components) to fit in with most other stuff. This is pretty inconsistently enforced though (sorry)

```suggestion
function getButtonText (rewardsEnabled: boolean,
```

**Review by fallaciousreasoning — APPROVED** [review]
LGTM, with a few (mostly optional) nits

**Comment by mkarolin** [comment]
Thanks @fallaciousreasoning, all great points. Addressed in [23d9e1a](https://github.com/brave/brave-core/pull/16895/commits/23d9e1ac2becc19cd7a9b3922597e1ce194a2b8b)

**Review by zenparsing — APPROVED** [review]

File: components/brave_new_tab_ui/containers/newTab/settings/icons/sponsoredImageInfo.tsx
Comment (petemill) [reply components/brave_new_tab_ui/containers/newTab/settings/icons/sponsoredImageInfo.tsx]: Semi-related question: If we had a lot of components using this pattern in our codebase, does it negatively affect performance for sessions which don't result in uses of those components. Instead of exporting functions that don't get called, there would instead be a lot of React.createClass calls during initial JS module parsing. A single (or a few hundred??) instances of this of course won't matter, but it would be interesting to know if this is a best practice and at what point it causes negative impact.

**Comment by mkarolin** [comment]
Squashed commits.

File: components/brave_new_tab_ui/containers/newTab/settings/sponsoredImagesToggle.tsx
Comment (petemill) [inline components/brave_new_tab_ui/containers/newTab/settings/sponsoredImagesToggle.tsx:65]: This won't show for non-mouse users. I think it could be pretty trivial to have the tooltip show when the icon is focused. Could you try?

File: components/brave_new_tab_ui/containers/newTab/settings/sponsoredImagesToggle.tsx
Comment (petemill) [inline components/brave_new_tab_ui/containers/newTab/settings/sponsoredImagesToggle.tsx:59]: In RTL UI (which you can force and still keep english via `chrome://flags`), this is making the text get clipped
image

This could get fixed by the following (as I can't think of a better way right now!):
```css
[dir=rtl] & {
  right: -200px;
}
```

Unfortunately that puts the arrow in the wrong place, so that also needs repositioning similarly.
image

**Review by petemill — CHANGES_REQUESTED** [review]
Just some issues with positioning in rtl and keyboard activation

File: components/brave_new_tab_ui/containers/newTab/settings/sponsoredImagesToggle.tsx
Comment (mkarolin) [reply components/brave_new_tab_ui/containers/newTab/settings/sponsoredImagesToggle.tsx:59]: @petemill `inset-inline-start` seems to work for both directions. WDYT?

Comment (mkarolin) [reply components/brave_new_tab_ui/containers/newTab/settings/sponsoredImagesToggle.tsx]: @petemill It also looks like toggles' bullet doesn't move correctly to the ON position with RTL set.

**Comment by mkarolin** [comment]
@petemill, could you, please, take another look. Thanks

**Comment by mkarolin** [comment]
Updated screenshots:

<table>
<tr><td>LTR</td><td>RTL</td></tr>
<tr><td></td><td>
</td></tr>
<tr><td></td><td>

</td></tr>
</table>

**Review by petemill — APPROVED** [review]
Thanks for addressing!

**Pull request merged**
